crypto/pkcs7/bio/md.c

```diff
@@ -0,0 +1,177 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+
+#include <errno.h>
+#include <openssl/buffer.h>
+#include <openssl/evp.h>
+#include <openssl/pkcs7.h>
+#include "../../internal.h"
+#include "../crypto/bio/internal.h"
+#include "../internal.h"
+
+// BIO_put and BIO_get both add to the digest, BIO_gets returns the digest
+static int md_write(BIO *h, char const *buf, int num);
+static int md_read(BIO *h, char *buf, int size);
+static int md_gets(BIO *h, char *str, int size);
+static long md_ctrl(BIO *h, int cmd, long arg1, void *arg2);
+static int md_new(BIO *h);
+static int md_free(BIO *data);
+
+
+static const BIO_METHOD methods_md = {
+    BIO_TYPE_MD, "message digest", md_write, md_read, /*puts*/ NULL,
+    md_gets,     md_ctrl,          md_new,   md_free, /*callback_ctrl*/ NULL,
+};
+const BIO_METHOD *BIO_f_md(void) { return &methods_md; }
+
+static int md_new(BIO *b) {
+  GUARD_PTR(b);
+  EVP_MD_CTX *ctx;
+
+  ctx = EVP_MD_CTX_new();
+  if (ctx == NULL) {
+    return 0;
+  }
+
+  BIO_set_data(b, ctx);
+
+  return 1;
+}
+
+static int md_free(BIO *b) {
+  GUARD_PTR(b);
+  EVP_MD_CTX_free(BIO_get_data(b));
+  BIO_set_data(b, NULL);
+  BIO_set_init(b, 0);
+
+  return 1;
+}
+
+static int md_read(BIO *b, char *out, int outl) {
+  GUARD_PTR(b);
+  GUARD_PTR(out);
+  int ret = 0;
+  EVP_MD_CTX *ctx;
+  BIO *next;
+
+  ctx = BIO_get_data(b);
+  next = BIO_next(b);
+
+  if ((ctx == NULL) || (next == NULL) || outl <= 0) {
+    return 0;
+  }
+
+  ret = BIO_read(next, out, outl);
+  if (BIO_get_init(b)) {
+    if (ret > 0) {
+      if (EVP_DigestUpdate(ctx, (unsigned char *)out, ret) <= 0) {
+        return -1;
+      }
+    }
+  }
+  BIO_clear_retry_flags(b);
+  BIO_copy_next_retry(b);
+  return ret;
+}
+
+static int md_write(BIO *b, const char *in, int inl) {
+  GUARD_PTR(b);
+  GUARD_PTR(in);
+  int ret = 0;
+  EVP_MD_CTX *ctx;
+  BIO *next;
+
+  if (inl <= 0) {
+    return 0;
+  }
+
+  ctx = BIO_get_data(b);
+  next = BIO_next(b);
+  if ((ctx != NULL) && (next != NULL)) {
+    ret = BIO_write(next, in, inl);
+  }
+
+  if (BIO_get_init(b)) {
+    if (ret > 0) {
+      if (!EVP_DigestUpdate(ctx, (const unsigned char *)in, ret)) {
+        BIO_clear_retry_flags(b);
+        return 0;
+      }
+    }
+  }
+  if (next != NULL) {
+    BIO_clear_retry_flags(b);
+    BIO_copy_next_retry(b);
+  }
+  return ret;
+}
+
+static long md_ctrl(BIO *b, int cmd, long num, void *ptr) {
+  GUARD_PTR(b);
+  EVP_MD_CTX *ctx, **pctx;
+  EVP_MD *md;
+  long ret = 1;
+  BIO *next;
+
+  ctx = BIO_get_data(b);
+  next = BIO_next(b);
+
+  switch (cmd) {
+    case BIO_CTRL_RESET:
+      if (BIO_get_init(b)) {
+        ret = EVP_DigestInit_ex(ctx, EVP_MD_CTX_md(ctx), NULL);
+      } else {
+        ret = 0;
+      }
+      if (ret > 0) {
+        ret = BIO_ctrl(next, cmd, num, ptr);
+      }
+      break;
+    case BIO_C_GET_MD_CTX:
+      pctx = ptr;
+      *pctx = ctx;
```

Comment:
From cross-referencing the [OpenSSL implementation](https://github.com/openssl/openssl/blob/OpenSSL_1_1_1-stable/crypto/evp/bio_md.c#L167C9-L167C21), does this need a call to `BIO_set_init`?

Reply:
I removed that intentionally. i don't think it makes much sense for a getter to set initialization.

Reply:
There's mention of why this is done in OpenSSL's documentation: https://docs.openssl.org/master/man3/BIO_f_md/#description

> Calling BIO_get_md_ctx() will return the context and initialize the BIO state. This allows applications to initialize the context externally if the standard calls such as BIO_set_md() are not sufficiently flexible.

I think the idea is that the caller can "get" the context, allowing it can perform whatever initialization it requires.  So the act of getting the context implies that the caller will ensure it's initialized.

Reply:
Ah, ok. That makes sense given that we otherwise rely on `BIO_set_md` to init the BIO state. will replace.